daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SValue.scala

```diff
@@ -151,9 +153,77 @@ object SValue {
       s"SPAP($prim, ${actuals.asScala.mkString("[", ",", "]")}, $arity)"
   }
 
+  /** We split SRecord (interface) from SRecordRep (implementation/representation)
+    *
+    * The interface supports creation from separate arrays of fields and values
+    * This is used throughout test code.
+    *
+    * The interface also supports (via unapply) the (legacy) reverse deconstruction.
+    * This is used by daml-script.
+    *
+    * The implementation/representation is via a scala Map.
+    * This representation is simple to manipulate (lookupField/updateField).
+    *
+    * This representation makes illegal cases unrepresentable -- i.e.
+    * - mismatched counts of fields/values
+    * - repeated field names
+    * And prevent brittle access to element values via indexing.
+    *
+    * Also note updateField has logarithmic complexity (where N is the number of fields)
+    * rather than the linear complexity that an array of element values would entail.
+    *
+    * The representation also includes an ordered field list.
+    * This is needed to support the legacy interface used by daml-script.
+    * And is also used when we convert the svalue back to a normalised LF value.
+    */
+
+  sealed abstract class SRecord extends SValue {
+    def id: Identifier
+    def fields: ImmArray[Name]
+    def values: util.ArrayList[SValue]
+    def lookupField(field: Name): SValue
+    def updateField(field: Name, value: SValue): SRecord
+  }
+
+  object SRecord {
+    def apply(id: Identifier, fields: ImmArray[Name], values: util.ArrayList[SValue]): SRecord = {
+      assert(fields.length == values.size)
+      val zipped = (fields.toSeq zip values.asScala).toList.reverse
+      SRecordRep(id, fields, zipped.toMap)
```

Comment:
Can we maybe consider some versions of this that assert information about field names/types?
Also, whats the intended way to work with maps in implementation code? Am I supposed to be building `SRecordRep`s?

Reply:
As we discussed, there is one place in daml-script code that could make use of a builder which takes a Map.